built_in_tasks/target_tracking_task.py

```diff
@@ -0,0 +1,494 @@
+    reward_time = traits.Float(.5, desc="Length of reward dispensation")
+    timeout_time = traits.Float(10, desc="Time allowed to go between targets")
+    max_distance_error = traits.Float(2, desc="Maximum deviation from the trajectory for reward (cm)")
+
+    def init(self):
+        self.trial_dtype = np.dtype([('trial', 'u4'), ('index', 'u4'), ('target', 'f8',(3,)), ('disturbance_path', 'f8',(3,)), ('is_disturbance', '?')])
+        super().init()
+
+    def _parse_next_trial(self):
+        '''Check that the generator has the required data'''
+        self.gen_indices, self.targs, self.disturbance_trial, self.disturbance_path = self.next_trial
+        #self.targs = np.concatenate((np.zeros(30),np.squeeze(self.targs,axis=0),np.zeros(60)))
+        self.targs = np.squeeze(self.targs,axis=0)
+        self.disturbance_path = np.squeeze(self.disturbance_path)
+
+
+        for i in range(len(self.disturbance_path)):
+            # Update the data sinks with trial information
+            self.trial_record['trial'] = self.calc_trial_num()
+            self.trial_record['index'] = self.gen_indices
+            self.trial_record['target'] = self.targs[i]
+            self.trial_record['disturbance_path'] = self.disturbance_path[i]
+            self.trial_record['is_disturbance'] = self.disturbance_trial
+            self.sinks.send("trials", self.trial_record)
+
+    def _start_wait(self):
+        # Call parent method to draw the next target capture sequence from the generator
+        super()._start_wait()
+
+        # number of times this sequence of targets has been attempted
+        self.tries = 0
+
+        # number of targets to be acquired in this trial
+        self.chain_length = len(self.targs)
+
+    def _start_target(self):
+        self.target_index += 1
+        self.frame_index = 0
+        self.total_distance_error = 0        
+
+    def _while_target(self):
+        # Calculate and sum distance between center of cursor and current target position
+        self.total_distance_error += self.test_in_target() 
+        
+        # Add Disturbance
+        self.plant_position.append(self._get_manual_position()[0])
+        self.disturbance_position = self.add_disturbance(self.plant_position[-1], self.plant_position[-1]-self.plant_position[-2], self.disturbance_path[self.frame_index],self.disturbance_path[self.frame_index-1])
+
+        # Move Target to next frame so it appears to be moving
+        self.update_frame()
+        
+        # Check if the trial is over and there are no more target frames to display
+        if self.frame_index >= np.shape(self.targs)[0]:
+            self.trial_timed_out = True
+    
+    def update_frame(self):
+        self.targets[self.target_index].move_to_position(np.array([-self.frame_index/3,0,0]))
+        if self.frame_index >= 30:
+            self.tracker.move_to_position(np.array([0,0,self.targs[self.frame_index-30][2]/4]))
+        self.targets[self.target_index].show()
```

Comment:
i don't understand this, could you make some comments to explain what the divide by 3 and the frame_index-30 are for?